gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -294,49 +328,145 @@ public void run() {
 
               private ByteString readRequestData() throws IOException {
-                // Mark the input stream in case this request fails so that read can be recovered
-                // from where it's marked.
-                pipeSource.mark(MAX_BYTES_PER_MESSAGE);
                 ByteString data =
                     ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
-
-                objectFinalized =
-                    data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.available() <= 0;
+                pipeSource.mark(1);
+                objectFinalized = data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.read() == -1;
+                pipeSource.reset();
                 return data;
               }
+
+              // Handles the case when a writeOffset of data read previously is being processed.
+              // This happens if a transient failure happens while uploading, and can be resumed by
+              // querying the current committed offset.
+              private InsertObjectRequest.Builder buildRequestFromCachedDataChunk(
+                  TreeMap<Long, ByteString> dataChunkMap) {
+                // Resume will only work if the first request builder in the cache carries an offset
+                // not greater than the current writeOffset.
+                InsertObjectRequest.Builder requestBuilder = null;
+                if (dataChunkMap.size() > 0 && dataChunkMap.firstKey() <= writeOffset) {
+                  long prevWriteOffset = 0;
+                  ByteString prevDataChunk = null;
+                  for (Map.Entry<Long, ByteString> entry : dataChunkMap.entrySet()) {
+                    // Take the previous entry if it's greater than current writeOffset.
+                    if (entry.getKey() == writeOffset) {
+                      Long curWriteOffset = entry.getKey();
+                      chunkData = entry.getValue();
+                      requestBuilder =
+                          buildRequestFromWriteOffsetAndDataChunk(curWriteOffset, entry.getValue());
+                      int count =
+                          retryMap.containsKey(curWriteOffset) ? retryMap.get(curWriteOffset) : 0;
+                      if (count > UPLOAD_RETRIES) {
+                        nonTransientError =
+                            new RuntimeException(
+                                String.format(
+                                    "Max retry attempts reached for uploadId %s at offset %d",
+                                    uploadId, curWriteOffset));
+                      } else {
+                        retryMap.put(curWriteOffset, count + 1);
+                      }
+                      break;
+                    } else if (entry.getKey() > writeOffset) {
+                      chunkData = prevDataChunk;
+                      requestBuilder =
+                          buildRequestFromWriteOffsetAndDataChunk(prevWriteOffset, prevDataChunk);
+                      int count =
```

Comment:
How about moving this check above the loop - it's loop-invariant, so doing that would make the loop shorter/more readable.

Reply:
The count is actually related to iteration itself. We have no idea upfront which <offset, datachunk> will be retired and which are not until we iterate and find the first entry that meed resume requirement.